Sc24 Updates

jolly-chen opened the pull request:
- Remove compilation of nonexistent directory `OneMKL_gemm`
- Modified the quick reference in the exercise source: mention device pointers instead of buffers. Also added an example of declaration of  `parallel_for` with and without command group handler 
- Modified the exercise solution to use USM device pointers 
- Modified the "Accessing Data With Accessors" slides to use USM device pointers and showcase accessing data with `id/item/nd_item`
- Changed word ordering on "Expressing parallelism" slide to make it easier to distinguish the differences between `id/item/nd_item`

hdelan commented on Code_Exercises/ND_Range_Kernel/solution.cpp:
```diff
@@ -29,22 +29,25 @@ TEST_CASE("range_kernel_with_item", "nd_range_kernel_solution") {
+    auto ptrB = sycl::malloc_device<int>(dataSize, gpuQueue);
+    auto ptrR = sycl::malloc_device<int>(dataSize, gpuQueue);
 
+    gpuQueue.memcpy(ptrA, a, sizeof(int) * dataSize).wait();
```
I don't really like the use of `.wait()` on every single async call. Would it be overkill to use `sycl::event`s to handle dependencies? It depends on the order that you will be presenting these on the day. Will synchronisation with events have been covered at this point? If not, then maybe some calls to `gpuQueue.wait()` at various points (between memcpys and kernel execution. `parallel_for(...).wait()` is OK IMO).

hdelan commented on Code_Exercises/ND_Range_Kernel/solution.cpp:
```diff
@@ -69,24 +72,26 @@ TEST_CASE("nd_range_kernel", "nd_range_kernel_solution") {
-      sycl::accessor accA{bufA, cgh, sycl::read_write};
-      sycl::accessor accB{bufB, cgh, sycl::read_write};
-      sycl::accessor accR{bufR, cgh, sycl::read_write};
+    gpuQueue.memcpy(ptrA, a, sizeof(int) * dataSize).wait();
```
Likewise

hdelan reviewed (commented):
Looks great Jolly thanks for the contribution! Would it make sense to do a `malloc_shared` instead of `malloc_device`? For the purpose of this exercise it might be nice to not need the memcpys?

jolly-chen commented:
Hi Hugh! Thanks for the review :smile: I changed the first solution to use `malloc_shared`, but I kept `malloc_device` for the second one (with fewer waits), because I think it could be nice to have an example of both types of allocations. 

Currently, we don't have `malloc_shared` on any of the slides, so I've added a slide on this to the introduction. Since this is your part, do you agree with the change @TApplencourt? I also noticed that the introduction slides about "SYCL Execution Model" (22-26) are repeated in ND_Range_Kernel. Maybe these could be removed from the introduction (and replaced with some slides on synchronization)?

TApplencourt commented:
Good idea! Sound good to me.

Agree about synchronization: 

In my own tutorial : `https://github.com/argonne-lcf/sycltrain/blob/master/9_sycl_of_hell/README.md`

I do like this

- Single task # use printf
-  Range
- ND_range
- Malloc shared # now using array
- Malloc device / memcopy
- In order queue #synchronization
- Out of Order queue # synchronization
- Reduction

hdelan commented on Lesson_Materials/Introduction_to_SYCL/index.html:
```diff
@@ -358,7 +358,26 @@
+					<div class="container" data-markdown>
+						* A USM shared allocation is performed by calling one of the `malloc_shared` functions.
+						* Shared allocations implicitly share data between the host and devices; the runtime and backends ensure the data is available when used.
+						* Performance hints may be specified by enqueueing `prefetch` operations
```
Well noted!

hdelan reviewed (approved):
LGTM. As long as the event stuff has been covered before the use of memcpy passing the event as an arg. Thank!

TApplencourt commented:
(merging so I can benefit from @jolly-chen  change in the introduction :) )
